climada/util/yearsets.py

```diff
@@ -0,0 +1,247 @@
+            (also for a different event_impacts object)
+        correction_fac : boolean
+            If True a correction factor is applied to the resulting annual_impacts. They are
+            scaled in such a way that the expected annual impact (eai) of the annual_impacts
+            equals the eai of the events_impacts
+
+    Returns:
+      annual_impacts : impact object
+          annual impacts for all sampled_years
+      sampling_dict : dict
+          the sampling dictionary containing two arrays:
+              selected_events (array) : sampled events (len: total amount of sampled events)
+              events_per_year (array) : events per sampled year
+          Can be used to re-create the exact same annual_impacts yearset
+      """
+
+
+    if not sampled_years and not sampling_dict:
+        sampled_years = np.arange(1, 1000+1).tolist()
+    elif isinstance(sampled_years, int):
+        sampled_years = np.arange(1, sampled_years+1).tolist()
+    elif not sampled_years:
+        sampled_years = np.arange(1, len(sampling_dict['selected_events'])+1).tolist()
+    elif len(sampled_years) != len(sampling_dict['events_per_year']):
+        LOGGER.info("The number of sampled_years and the length of the list of events_per_year "
+                    "in the sampling_dict differ. The number of years contained in the "
+                    "sampling_dict are used as number of sampled_years.")
+        sampled_years = np.arange(1, len(sampling_dict['selected_events'])+1).tolist()
+
+    if sampling_dict and (
+            sum(sampling_dict['events_per_year']) != len(sampling_dict['selected_events'])):
+        raise ValueError("The sampling dictionary is faulty: the sum of selected events "
+                         "does not correspond to the number of selected events.")
+
+    n_sampled_years = len(sampled_years)
+
+    if len(np.unique(event_impacts.frequency)) > 1:
+        LOGGER.warning("The frequencies of the single events in the given event_impacts "
+                       "differ among each other. Please beware that this will influence "
+                       "the resulting annual_impacts as the events are sampled uniformaly "
+                       "and different frequencies are (not yet) taken into account.")
+
+
+    # INITIALISATION
+    # to do: add a flag to distinguish annual_impacts from other impact class objects
+    # artificially generate a generic  annual impact set (by sampling the event impact set)
+    annual_impacts = copy.deepcopy(event_impacts)
+    annual_impacts.event_id = [] # to indicate not by event any more
+    annual_impacts.event_id = np.arange(1, n_sampled_years+1)
+    annual_impacts.frequency = [] # init, see below
+    annual_impacts.at_event = np.zeros([1, n_sampled_years+1])
+
+    #create sampling dictionary if not given as input
+    if not sampling_dict:
+        n_annual_events = np.sum(event_impacts.frequency)
```

Comment:
@chahank I'm questioning this computation - would it not make more sense to calculate the average number of annual events as np.mean(1/event_impacts.frequency)?

Reply:
I think your formula is not correct. Suppose I have one event which happens `1/100` years, thus `event_impacts.frequency = [1/100]`. Then `np.mean(1/event_impacts.frequency) = 100` instead of `1/100`.

Reply:
you're right, I got confused with the frequencies yesterday. I think the formula is correct as is :).